compiler/rustc_middle/src/hir/map/mod.rs

```diff
@@ -1322,7 +1322,9 @@ impl<'hir> Visitor<'hir> for ItemCollector<'hir> {
 
     fn visit_item(&mut self, item: &'hir Item<'hir>) {
         if associated_body(Node::Item(item)).is_some() {
-            self.body_owners.push(item.owner_id.def_id);
+            if !self.tcx.has_attr(item.owner_id.def_id, sym::rustc_intrinsic_must_be_overridden) {
```

Comment:
Why was this necessary? I don't like it, can this be avoided by making the bodies of must_be_overriden intrinsics something else (like another intrinsic or intrinsics::unreachable^^)?

Reply:
Basically this avoids ever invoking typeck on those dummy bodies. We have various places that iterate over all body owners, some of them will try to invoke mir_const_qualifs, borrowck or just optimized_mir directly. I forget why we didn't need this check before this PR. const_eval_select was somehow special here...

Reply:
If we used `intrinsics::unreachable` (or maybe some short wrapper `must_override_body` function instead, would all of these problems go away? Hacking around not wanting to invoke typeck seems fragile

Reply:
Well, I don't want to return a value from `optimized_mir`, because that body could then end up in codegen somehow. So I added an assertion in mir_built to ensure we never build MIR for these intrinsics. That assert kept getting hit in the first commit, precisely because we assumed that any body owner will have a MIR body. Removing these intrinsics from MIR body owners sidesteps this issue in a fairly noninvasive way. It is fishy, I agree, but I also don't think it's worth adding body-less functions to HIR, AST and the parser, just for these intrinsics.

Reply:
Hm, how about making them just recurse? Typeck and MIR building etc should like that. It doesn't *guarantee* that the body won't be codegened, but I think that's okay.

Reply:
I just fear we'll screw up somewhere and end up actually invoking that body. An early impl of `rustc_intrinsic_must_be_overridden` had that issue for some cases. I guess I could codegen them all into an abort so it's obvious when it happens. It's very easy to debug with the assert in mir_built, but one can always add it again, and we can still avoid putting that MIR into metadata, causing ICEs in all crates but codegen.

Reply:
This has probably come up before, but what about making them `extern "Rust"` with a `#[rustc_intrinsic]` attribute?

Reply:
The reason we went away from foreign functions wasn't just the weird ABI, but also the problem that we can't specify safety or constness on foreign functions